Refine the grouped InteractionRegions' geometry

etiennesegonzac opened the pull request:
#### ac3d70bd04ef227fb925de4019b01348654e0da0
<pre>
Refine the grouped InteractionRegions&apos; geometry
<a href="https://bugs.webkit.org/show_bug.cgi?id=255654">https://bugs.webkit.org/show_bug.cgi?id=255654</a>
&lt;rdar://107760442&gt;

Reviewed by Tim Horton.

Use masked corners to &quot;shrink wrap&quot; grouped InteractionRegions. And
make them fit the element&apos;s visual style better.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Lookup the `borderRadius` on the region&apos;s renderer and not the renderer of
the matched element.
Only apply inline bounds inflation and minimum border radius to
renderers without a visual edge.

* Source/WebCore/rendering/EventRegion.cpp:
* Source/WebCore/rendering/EventRegion.h:
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
Introduce a new method to edit the masked corners on grouped
InteractionRegions and make them more visually pleasing.

* LayoutTests/interaction-region/overlap-same-group-expected.txt:
* LayoutTests/interaction-region/overlap-same-group.html:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link.html:
Test updates covering the masked corners changes.

* LayoutTests/interaction-region/border-radii-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/overlap-expected.txt:
* LayoutTests/interaction-region/tiny-regions-expected.txt:
Test updates now showing the element&apos;s real border radius.

* LayoutTests/interaction-region/layer-tree-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
Test updates.

Canonical link: <a href="https://commits.webkit.org/263337@main">https://commits.webkit.org/263337@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d798b77823d454922a49ed9ef42906d3080bd1e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3710 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3795 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3902 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5138 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3992 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3683 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3872 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3800 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3214 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3754 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3985 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3345 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4971 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1491 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3314 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3379 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3300 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3373 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4743 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3762 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3049 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3294 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3314 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1054 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3330 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3571 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -200,13 +200,10 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
     OptionSet<InteractionRegion::CornerMask> maskedCorners;
 
-    if (auto* renderBox = dynamicDowncast<RenderBox>(renderer)) {
+    if (auto* renderBox = dynamicDowncast<RenderBox>(regionRenderer)) {
```
Do you think this is strictly a win? Should we be max() or min()ing them or something?

etiennesegonzac replied:
yep, especially with the new, higher, minimum border radius

etiennesegonzac committed:
Refine the grouped InteractionRegions' geometry
https://bugs.webkit.org/show_bug.cgi?id=255654
<rdar://107760442>

Reviewed by Tim Horton.

Use masked corners to "shrink wrap" grouped InteractionRegions. And
make them fit the element's visual style better.

* Source/WebCore/page/InteractionRegion.cpp:
(WebCore::interactionRegionForRenderedRegion):
Lookup the `borderRadius` on the region's renderer and not the renderer of
the matched element.
Only apply inline bounds inflation and minimum border radius to
renderers without a visual edge.

* Source/WebCore/rendering/EventRegion.cpp:
* Source/WebCore/rendering/EventRegion.h:
(WebCore::EventRegionContext::shrinkWrapInteractionRegions):
(WebCore::EventRegionContext::copyInteractionRegionsToEventRegion):
Introduce a new method to edit the masked corners on grouped
InteractionRegions and make them more visually pleasing.

* LayoutTests/interaction-region/overlap-same-group-expected.txt:
* LayoutTests/interaction-region/overlap-same-group.html:
* LayoutTests/interaction-region/wrapped-inline-link-expected.txt:
* LayoutTests/interaction-region/wrapped-inline-link.html:
Test updates covering the masked corners changes.

* LayoutTests/interaction-region/border-radii-expected.txt:
* LayoutTests/interaction-region/consolidated-nested-regions-expected.txt:
* LayoutTests/interaction-region/overlap-expected.txt:
* LayoutTests/interaction-region/tiny-regions-expected.txt:
Test updates now showing the element's real border radius.

* LayoutTests/interaction-region/layer-tree-expected.txt:
* LayoutTests/interaction-region/paused-video-regions-expected.txt:
Test updates.

Canonical link: https://commits.webkit.org/263337@main
